### PR TITLE
Feature/add client metadata

### PR DESCRIPTION
**Pull request by pike1212**
## Related issue

#1594

## Proposed changes

Adds metadata field to oauth client

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

**Comment by pike1212** [comment]
`make test` works for me locally, I did run `make sdk` and commit those changes, but not sure why these tests are failing

File: client/manager_sql.go [hunk 1]
```diff
@@ -179,10 +186,21 @@ func sqlDataFromClient(d *Client) (*sqlData, error) {
...
 }
 
 func (d *sqlData) ToClient() (*Client, error) {
+	var metadata map[string]interface{}
+
+	if d.Metadata == "" {
+		d.Metadata = "{}"
+	}
+
+	if err := json.Unmarshal([]byte(d.Metadata), &metadata); err != nil {
```
Comment (aeneasr) [inline client/manager_sql.go]:
```suggestion
	if err := json.Unmarshal([]byte(stringsx.Coalesce(d.Metadata, "{}")), &metadata); err != nil {
```

File: client/manager_sql.go [hunk 2]
```diff
@@ -179,10 +186,21 @@ func sqlDataFromClient(d *Client) (*sqlData, error) {
...
 		BackChannelLogoutURI:              d.BackChannelLogoutURI,
 		BackChannelLogoutSessionRequired:  d.BackChannelLogoutSessionRequired,
+		Metadata:                          string(metadata),
 	}, nil
 }
 
 func (d *sqlData) ToClient() (*Client, error) {
+	var metadata map[string]interface{}
+
+	if d.Metadata == "" {
```
Comment (aeneasr) [inline client/manager_sql.go]: Can be omitted if change below is applied:

```suggestion
```

File: client/manager_sql.go [hunk 3]
```diff
@@ -213,7 +231,8 @@ func (d *sqlData) ToClient() (*Client, error) {
 		FrontChannelLogoutSessionRequired: d.FrontChannelLogoutSessionRequired,
 		PostLogoutRedirectURIs:            stringsx.Splitx(d.PostLogoutRedirectURIs, "|"),
 		BackChannelLogoutURI:              d.BackChannelLogoutURI,
-		BackChannelLogoutSessionRequired:  d.BackChannelLogoutSessionRequired,
+		BackChannelLogoutSessionRequired:  d.BackChannelLogoutSessionRequired, //
+		Metadata:                          metadata,
```
Comment (aeneasr) [inline client/manager_sql.go]: If `metadata` is `json.RawMessage` there would be no need for decoding as we could do:

```suggestion
		Metadata:                          []byte(metadata),
```

Then, all previous comments (Coalesce etc) would no longer be required.

File: client/x_manager_sql_migrations_test.go [hunk 4]
```diff
@@ -34,10 +34,10 @@ import (
 )
 
 func CleanTestDB(t *testing.T, db *sqlx.DB) {
-	_, err := db.Exec("DROP TABLE IF EXISTS hydra_client_migration")
-	t.Logf("Unable to execute clean up query: %s", err)
-	_, err = db.Exec("DROP TABLE IF EXISTS hydra_client")
-	t.Logf("Unable to execute clean up query: %s", err)
+	// _, err := db.Exec("DROP TABLE IF EXISTS hydra_client_migration")
```
[the remark this answers is not in the document]
Comment (pike1212) [reply client/x_manager_sql_migrations_test.go]: reverted

File: client/x_manager_sql_migrations_test.go [hunk 5]
```diff
@@ -55,7 +55,7 @@ func TestXXMigrations(t *testing.T) {
 		CleanTestDB, CleanTestDB,
 		func(t *testing.T, dbName string, db *sqlx.DB, _, step, steps int) {
 			if dbName == "cockroach" {
-				step += 12
```
[the remark this answers is not in the document]
Comment (pike1212) [reply client/x_manager_sql_migrations_test.go]: reverted

File: client/migrations/sql/postgres/14.sql [hunk 6]
```diff
@@ -0,0 +1,5 @@
+-- +migrate Up
+ALTER TABLE hydra_client ADD metadata TEXT NOT NULL DEFAULT '{}';
```
[the remark this answers is not in the document]
Comment (pike1212) [reply client/migrations/sql/postgres/14.sql]: changed to similar pattern as mysql (add column, set values, set not null)

File: client/manager_sql.go (hunk 3 above)
Comment (pike1212) [reply client/manager_sql.go]: I made this change

File: client/manager_sql.go (hunk 2 above)
Comment (pike1212) [reply client/manager_sql.go]: Removed this code

File: client/manager_sql.go (hunk 1 above)
Comment (pike1212) [reply client/manager_sql.go]: Removed this code

File: client/migrations/sql/cockroach/14.sql [hunk 7]
```diff
@@ -32,6 +32,7 @@ CREATE TABLE IF NOT EXISTS hydra_client (
 	post_logout_redirect_uris TEXT NOT NULL DEFAULT '',
 	backchannel_logout_uri TEXT NOT NULL DEFAULT '',
 	backchannel_logout_session_required BOOL NOT NULL DEFAULT FALSE,
+	metadata TEXT NOT NULL DEFAULT '{}',
```
[the remark this answers is not in the document]
Comment (pike1212) [reply client/migrations/sql/cockroach/14.sql]: A little unsure what needs done here, I added the 13.sql back and added a 14.sql with an alter statement, but I think the tests are still failing

**Comment by aeneasr** [comment]
Please run [`make sqlbin`](https://github.com/ory/hydra/blob/master/Makefile#L76-L82), then tests should pass.

**Comment by aeneasr** [comment]
Thank you for your contribution!

**Pull request merged**